ALM.Task_1

volodymyr921 opened the pull request.

volodymyr921 committed:
init

volodymyr921 committed:
README.md

AndryLat commented on .idea/vcs.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="VcsDirectoryMappings">
-    <mapping directory="$PROJECT_DIR$" vcs="Git" />
```
Any files related with your LOCAL env should not be in repository. Here you should keep only common files for whole team and only related with project.

AndryLat replied:
It means that IDEA files, Linux\windows files (your custom scripts for debugging, but not for production), local settings file - should be absent in repo. 
Common files for team mean files which are needed for any team mate for running \ developing project on his local env.

volodymyr921 replied:
Fixed here: 13246fb

AndryLat commented on src/main/java/CreditCardNumber.java:
```diff
@@ -0,0 +1,9 @@
+public class CreditCardNumber {
```
Where is packaging? 

https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html#:~:text=Naming%20Conventions,a%20programmer%20at%20example.com%20.

It is very important for each codebase.

AndryLat replied:
When you fixed something, it is better to show which commit fixes it.

Like:

Fixed here: 874177832ab23a5feb5c9480fc4c977ab8e9b1c2

@volodymyr921

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/UserCardNumber.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * This class, UserCardNumber, is designed to obtain a card number from the user for validation.
+ */
+public class UserCardNumber {
```
Name of the class say nothing about the main idea. during the call I will explain how it should be.

AndryLat replied:
ConsoleUtils.readCardNumber()
Utils.readcardNumberFromConsole();
Utils.readcardNumberFromFile();

UserDAO 
-- fetch
-- save
-- delete
-- update

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/UserCardNumber.java:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * This method displays a message to the user asking for a card number.
+     */
+    public void promptToTheUser() {
```
Why is the method public? Do you really want to provide access to it?

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+     * Validates the provided credit card number for length, digit-only content, and determines the payment system.
+     * Records any encountered errors during validation.
+     */
+    public void validate() {
```
"validate" means validation. But you confuse reader. The method make transformation, validation and payment system determination.

AndryLat replied:
```
public void validate() {
  validation(..);
  determination(..);
  printResult(..);
}

```

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+     * Records any encountered errors during validation.
+     */
+    public void validate() {
+        // Remove any white spaces from the card number
```
Do you really think that each line of production code is covered by comments? It makes code unreadable. The best code is with the minimum amount of comments. You should provide comments only in really tricky cases.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+     */
+    public void validate() {
+        // Remove any white spaces from the card number
+        cardNumber = cardNumber.replaceAll("\\s", "");
```
"\\s" - should be a constant. When it is constant, you can give clear name which describes idea of constant.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+        cardNumber = cardNumber.replaceAll("\\s", "");
+
+        // Check if the length of the card number is correct
+        if (cardNumber.length() != LENGTH_CARD_NUMBER) errors.add(ERROR_LENGTH_CARD);
```
don't use one line statements. It reduces readability of code.
Also always use "{}" for each operator to avoid potential bugs.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+        if (cardNumber.length() != LENGTH_CARD_NUMBER) errors.add(ERROR_LENGTH_CARD);
+
+        // Check if the card number contains only digits
+        int i = 0;
```
You can split one big method to few small and it will be more readable, supportable and clearer

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+        // If payment system couldn't be determined, add an error message
+        if (paymentSystem == null) errors.add(ERROR_CANNOT_DETERMINE_PAYMENT_SYSTEM);
+
+        displayTheResult();
```
Imagine case, that tomorrow you will need to use the same functionality for REST service. What will you do? Copy paste? Let's discuss it during the call.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @return The PaymentSystem enum representing the detected payment system, or null if not determined.
+     */
+    public PaymentSystem detectPaymentSystem() {
```
Class has name "ValidationCreditCardNumber" and do you really need to provide the method as API of the class?

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+                if (cardNumber.startsWith(String.valueOf(prefix))) return paymentSystem;
+            }
+        }
+        return null;
```
there are few best practices to provide null safety.
The simples one is using Optional.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/test/java/CreditCardValidateTest.java:
```diff
@@ -0,0 +1,52 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class CreditCardValidatorTest {
```
I see no validation of validation mechanism: exactly set of errors.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/PaymentSystem.java:
```diff
@@ -0,0 +1,18 @@
+        this.prefix = prefix;
+    }
+
+    public int[] getPrefix() {
```
Will I receive one prefix or few prefixes ? Name is confusing reader

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/main/java/ValidationCreditCardNumber.java:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * Displays the validation result: whether the card is valid or invalid, along with any encountered errors.
+     */
+    public void displayTheResult() {
```
articles should be avoided for using them in names.

volodymyr921 replied:
Fixed here: 8741778

AndryLat commented on src/test/java/CreditCardValidateTest.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Test
+    public void testLengthCard() {
```
shouldReturnNullWhenCardHasWrongLength

AndryLat replied:
shouldThrowWrongCardNumberException_WhenCardLengthIsWrong
shouldReturnMastercardPaymentSystem_WhenCardNumberIsCorrectAndStartsWith53
shouldReturnMastercardPaymentSystem_WhenCardNumberIsCorrectAndStartsWith54
shouldReturnMastercardPaymentSystem_WhenCardNumberIsCorrectAndStartsWith55

```
@ParanmetrizedTest
@ValueSource(strings= {"53..123", "54..123"})
void shouldReturnMastercardPaymentSystem_WhenCardNumberIsCorrect() {
...
}
```

volodymyr921 replied:
Fixed here: 8741778

volodymyr921 committed:
The structure of the task was reworked, the Lunр algorithm was refined, and the testing was completely reworked

volodymyr921 committed:
Merge remote-tracking branch 'origin/task' into task

volodymyr921 committed:
Delete .idea directory

volodymyr921 committed:
Update README.md

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/enums/Errors.java:
```diff
@@ -0,0 +1,17 @@
+package dev.andrylat.vomelianchuk.CardNumberOperations.enums;
+
+public enum Errors {
+    ERROR_LENGTH_CARD("Length should be 16 symbols"),
```
We are in the class "Errors" and it is default that all enums are error. You don't need to duplicate it.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/enums/PaymentSystem.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public static Optional<PaymentSystem> detectPaymentSystem(String cardNumber) {
+        if (cardNumber == null) return Optional.empty();
```
we discussed last time that you should always use {} for each operator like if, while, for, etc.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/ConsoleUtils.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.Scanner;
+
+public class ConsoleUtils {
```
Imagine situation that your project lives during last 10 years and this class contains 1000 different methods related with "Console", because it is "utils".

You should always try to make specific meaning for class and don't give possibility to another guys to put here some another methods. 

You can use name like "UserConsoleIO" and it will be enough. Utils here is not appropriate. Utils we use only for classes which are very common. In the best application these classes should be absent./

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+    private static final String HIATUS_SEPARATOR = "\\s";
+    private static final int TEN = 10;
+
+    private String cardNumber;
```
During implementation, the best functionality is when you call method, provide everything it needs and receive result. I mean that it is good to avoid unnecessary states. 
You should not use them there where we can avoid them.

Here you can create one public method and put there everything and receive result. 

Let's think, how it can be achieved.
What is needed as input? only card number. It means that you can create method "operate" and as parameter you can put card number.
As a result it can be payment system. Because you definitely need it. 
```
public PaymentSystem determinePaymentSystem(String cardNumber)
```

And you can have reasonable questions: what about errors? 
We discussed that you can throw exception if validation is failed. And don't forget that exception is a class and it means that you can store there aby data you want. 
So, put there list of errors. 
you should have something like 
```
class YourException extends RuntimeException {
  private List<String> errorMessages = ...;
  public YourException(String message, List<String> errorMessages) {
  ...
  }
  ...
}
```
And you have very easy use cases:
- if everything is ok, you will receive payment system.
- if no, you will catch exception with all errors.

And the most important part! 
If you make as I said, your code will be safe thread and it can work in multythreading env very good without any needs to create new object each time.
It is my main goal to you to explain it.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+        this.cardNumber = cardNumber;
+    }
+
+    public String operate() {
```
It is the worth way you could do.

Your "String" contains ANYTHING! and it is not logical that the method will return error. name doesn't say about it. I can think that as a result I can receive name of payment system, or error, or something else.

AndryLat replied:
and as I see it is done as I said. Why you return payment system as string? The code is not reusable in this case. and nobody knows when there is error in the returning value and when there is actual result. 

Don't think about code as for one place using. Always try to make maximum reusable code if it consumes not much time and effort.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+                determine();
+            }
+            return messageResult();
+        } catch (NullPointerException exception) {
```
Catching NPE is not good practice because it is indicator that your code is bad. Good code should avoid any NPE.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private String transform() throws NullPointerException {
+        return cardNumber.replaceAll(HIATUS_SEPARATOR, "");
```
You must make validation for null BEFORE doing something.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+
+    private String cardNumber;
+    private PaymentSystem paymentSystem;
+    private Errors error;
```
Only one error can be ? There are few errors for the same input in the example.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+        return sum;
+    }
+
+    private int calculateDigit(char digitChar) {
```
Functionality of Luhn algorithm can be moved to separate class because it is complex validation.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+        return (number >= TEN) ? (number % TEN) + 1 : number;
+    }
+
+    private void determine() {
```
If I send you this exact line, will you understand what is going there? I don't think so. But in good code, it should be.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumberProcess.java:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    private String messageResult() {
```
Method is an action. And name of method should contains action. Exception is only for data classes.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/test/java/CardNumberCorrectValidationTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Test
+    public void shouldDescriptionDinersClubPaymentSystem_WhenCardNumberIsCorrectAndStarsWith36() {
```
You can use parametric tests. In this case it is very good to apply them there. And yes, I know that in the ticket there is another information:) You should not trust tickets 100%.

volodymyr921 replied:
Fixed here: 1783555

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/CardNumberOperations/CardNumber.java:
```diff
@@ -0,0 +1,12 @@
+package dev.andrylat.vomelianchuk.CardNumberOperations;
```
You missed package for project. Because one student will have multiple projects.

volodymyr921 replied:
Fixed here: 1783555

volodymyr921 committed:
Reworked testing, included parameterized tests;
Moved Luhn algorithm to a separate class and reworked it for better clarity;
Changed the `determinePaymentSystem` method to either return the payment system or throw an exception with a list of errors;
And in general, I made several minor improvements to the code.

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumber.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.List;
+
+public class CardNumber {
```
Now, everything is much more better. I like it.

but solution is still not production like.

According to SOLID, your class should have only one responsibility. Here I see two of them:
- run Java application (main())
- validate card.

It is not correct. This class should be responsible only for running Java application. What does it mean? It means that only running related things should be here. What are they? 
 - Work with arguments
 - Main application flow (next task will be about adding mortgage calculator and selecting correct UI should be here)
 - etc...
 
 My main ideas is about moving all related stuff with validation to a separate class.

AndryLat replied:
the good solution should be like:
```
main() {
  UserConsoleIO  userConsoleIO = new UserConsoleIO(System.in, System.out);
  var cardNumberAsString = userConsoleIO.readCardNumber();
  try {
    new CardNumbervalidator().validate(cardNumberAsString);
    PaymentSystem paymentSystem = PaymentSystem.of(cardNumberAsString);
    userConsoleIO.printPaymentSytem(paymentSystem);
  } catch ( CardValidationException | WrongPaymentSystemException ex) {
    userConsoleIO.printErrors(ex);
  }
}

AndryLat replied:
Homework: why I use here "new UserConsoleIO(System.in, System.out)" ? What is the purpose to inject streams?

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on pom.xml:
```diff
@@ -6,4 +6,12 @@
   <version>1.0-SNAPSHOT</version>
   <name>Archetype - ALM.Task_1</name>
   <url>http://maven.apache.org</url>
+  <dependencies>
```
You can use Java 17, one of the most popular now. Just notice

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumber.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private static String getDescriptionPaymentSystem(PaymentSystem paymentSystem) {
+        return "\nCard is valid. Payment System is \"" + paymentSystem + "\"";
```
You can use formatted() method if you use Java 17. Also, UserConsoleIO class is responsible for IO. ONLY it.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumber.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private static String getDescriptionErrors(ExceptionValidation exceptionValidation) {
+        StringBuilder descriptionErrors = new StringBuilder("\nCard number is invalid.\nError:\n");
```
Statements like here can be moved to contants. Here it is optional, in another places (if it is regular expression pattern, etc) must be a constant.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/ExceptionValidation.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.List;
+
+public class ExceptionValidation extends RuntimeException {
```
Exceptions always should have "Exception" at the end of name. It is java code convention.
ValidationException is very generic name. CardNumberValidationException is better.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/ExceptionValidation.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+
+public class ExceptionValidation extends RuntimeException {
+    public static final String ERROR_NULL_CARD_NUMBER = "Error receiving null object instead of card number";
```
create these constants at the places where you use it. Because in a medium project the class will contain hundreds of contants, it will be awful for reading.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/LuhnAlgorithm.java:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    private boolean isTwoDigitNumber(int number) {
```
-100 will return true in this case. I think it is totally wrong.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/LuhnAlgorithm.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public boolean checkDigit(ExceptionValidation errors) throws ExceptionValidation {
+        if (cardNumber.isEmpty()) {
```
We can expect here that card number is valid. We don't need to duplicate validations everywhere.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/LuhnAlgorithm.java:
```diff
@@ -0,0 +1,75 @@
+package dev.andrylat.vomelianchuk.card_number_operations;
+
+public class LuhnAlgorithm {
+    private final String cardNumber;
```
Do you really need this state? Why you can't create method which receives card number as input and boolean as result (or void and throws exception)?

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/LuhnAlgorithm.java:
```diff
@@ -0,0 +1,75 @@
+        this.cardNumber = cardNumber;
+    }
+
+    public boolean checkDigit(ExceptionValidation errors) throws ExceptionValidation {
```
You use runtime unchecked exception, you don't need to use throws.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/LuhnAlgorithm.java:
```diff
@@ -0,0 +1,75 @@
+        int lastDigit = calculateLastDigit(firstNumber, secondNumber);
+
+        if (lastDigit == Character.getNumericValue(cardNumber.charAt(cardNumber.length() - 1))) {
+            return true;
```
the code will never return false. What is the purpose of having return value?

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumberProcess.java:
```diff
@@ -0,0 +1,66 @@
+    private static final String HIATUS_SEPARATOR = "\\s";
+
+    private PaymentSystem paymentSystem;
+    private ExceptionValidation errors = new ExceptionValidation(new ArrayList<>());
```
it is not good to save expection and use it as container. You can have list here and when you throw wxception, you put list there.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumberProcess.java:
```diff
@@ -0,0 +1,66 @@
+            if (validate(cardNumber)) {
+                determine(cardNumber);
+            } else {
+                throw new ExceptionValidation(errors.getErrorsMessage());
```
"errors" is expcetion, you can throw just errors. You don;'t need to create new one. But I described what is better solution.

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumberProcess.java:
```diff
@@ -0,0 +1,66 @@
+
+    private boolean checkLengthCardNumber(String cardNumber) {
+        if (cardNumber.length() != LENGTH_CARD_NUMBER) {
+            errors = new ExceptionValidation(ExceptionValidation.ERROR_LENGTH_CARD, errors.getErrorsMessage());
```
it can be rewrited as "errors.add(ExceptionValidation.ERROR_LENGTH_CARD)" when errors is just list.  it will be better and easier to understand

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/PaymentSystem.java:
```diff
@@ -0,0 +1,36 @@
+        return prefixes;
+    }
+
+    public static Optional<PaymentSystem> detectPaymentSystem(String cardNumber) {
```
or you can name this method as "of()"

volodymyr921 replied:
Fixed here: c79cac4

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/PaymentSystem.java:
```diff
@@ -0,0 +1,36 @@
+                }
+            }
+        }
+        return Optional.empty();
```
using optinal here is also very good option.

volodymyr921 committed:
Reworked the CardNumber main class to do one action, reworked some methods in the LuhnAlgorith class, added another error type

volodymyr921 committed:
Update README.md

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumber.java:
```diff
@@ -0,0 +1,17 @@
+                PaymentSystem paymentSystem = PaymentSystem.determine(cardNumberAsString);
+                UserConsoleIO.printPaymentSystem(paymentSystem);
+            } catch (CardValidationException | WrongPaymentSystemException e) {
+                UserConsoleIO.printErrors(e);
```
Why do you use static context if you have variable. this small mistake can invoke huge errors. 

What do I mean?
Now you created object userConsoleIO and that objects contains streams for input \ output. And, for example, one day I want to forward results to the stardard IO (console), another day - to the files.
And there is huge error will be: when I create userConsoleIO and use file's streams and I receive errors, I will receive list of errors in the CONSOLE, not file, which I defined!

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumberValidator.java:
```diff
@@ -0,0 +1,51 @@
+
+    private final List<String> errors = new ArrayList<>();
+
+    public void validate(String cardNumber) throws CardValidationException {
```
Nope, nope, nope! Very big mistake.

Or you use state in your class and you inject all necessary data to the constructor (as it was as I remember), and after it you call method validate(), getErrors(), etc.

OR you avoid using state and it means that you method receive all necessary information and return result (or return results, or throw exception if needed). BUT in this case YOU SHOULD NOT USE STATE in the class (line 13), because your code is not thread safe and you always must remember about cleanup. Our classes must be as simple as possible. I don't need to think more that it is needed during usage of class.

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumberValidator.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    private String transform(String cardNumber) {
```
minor: transform is very generaic name. Better to use something like "removeSeparators"

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardNumberValidator.java:
```diff
@@ -0,0 +1,51 @@
+
+    private void checkSymbolsCardNumber(String cardNumber) {
+        int i = 0;
+        while (i < cardNumber.length()) {
```
FOR is better for this case. It is much more easier to read. 95% time we use FOR operator.

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardValidationException.java:
```diff
@@ -0,0 +1,24 @@
+package dev.andrylat.vomelianchuk.card_number_operations;
```
You put everything in the same package. But you have absolutely different pieces of functionality and you can split them and put in the different packages.

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/CardValidationException.java:
```diff
@@ -0,0 +1,24 @@
+        return errorsMessage;
+    }
+
+    public String getErrorsMessagesInLine() {
```
Exception should not have responsibility for formatting messages. It is responsibility of class who uses it.

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/LuhnAlgorithm.java:
```diff
@@ -0,0 +1,72 @@
+
+
+public class LuhnAlgorithm {
+    private static final String ERROR_CANNOT_DETERMINE_PAYMENT_SYSTEM = "Payment System can't be determine";
```
Not clear: what is relation between control sum and payment system?

volodymyr921 replied:
Fixed here: 7a87509

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/PaymentSystem.java:
```diff
@@ -0,0 +1,47 @@
+        return prefixes;
+    }
+
+    public static PaymentSystem determine(String cardNumber) throws WrongPaymentSystemException {
```
throws WrongPaymentSystemException is not needed.

I will explain why:

Imagine situation, that this class is used in 46 different locations. and IN EACH LOCATION they have try-catch block to handle the exception. Everything is ok.
BUT: tomorrow you need to introduce NEW exception and add it to the throws statement. Could you imagine scope of changes? It is huge! Huge regression, huge potential chance to have a bug, etc. Don't use checked exceptions and throws statement.

volodymyr921 replied:
Fixed here: 7a87509

volodymyr921 committed:
Fixed bugs throughout the code and adjusted testing

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/consoleutils/UserConsoleIO.java:
```diff
@@ -0,0 +1,72 @@
+    private static final String INVALID_CARD_TEXT_BEGIN =  "\nCard number is invalid.\nError:\n";
+    private static final String INVALID_CARD_TEXT_SEPARATOR =  "\n";
+
+    private static InputStream input;
```
Why static? 

Static means that you can have only one instance of the object in whole application.

AndryLat commented on src/main/java/dev/andrylat/vomelianchuk/card_number_operations/validation/CardNumberValidator.java:
```diff
@@ -0,0 +1,54 @@
+    private static final String ERROR_LENGTH_CARD = "Length should be 16 symbols";
+    private static final String ERROR_ONLY_DIGITS = "Number should contain only digits";
+
+    public PaymentSystem validate(String cardNumber) throws CardValidationException {
```
We discussed that we don't use throws:)

AndryLat commented:
You can fix my comments and merge. I think, excellent job is done!

volodymyr921 committed:
Fixed last bug
